source/common/common/compiled_string_map.h

```diff
@@ -0,0 +1,235 @@
+      range_start = range_end;
+    }
+  }
+
+private:
+  /**
+   * Details of a node branch point; the index into the string at which
+   * characters should be looked up, the lowest valued character in the
+   * branch, the highest valued character in the branch, and how many
+   * branches there are.
+   */
+  struct IndexSplitInfo {
+    uint8_t index_, min_, max_, count_;
+    size_t size() { return max_ - min_ + 1; }
+    size_t offsetOf(uint8_t c) { return c - min_; }
+  };
+
+  /**
+   * @param node_contents the key-value pairs to be branched upon.
+   * @return details of the index on which the node should branch
+   *         - the index which produces the most child branches.
+   */
+  static IndexSplitInfo findBestSplitPoint(const std::vector<KV>& node_contents) {
+    IndexSplitInfo best{0, 0, 0, 0};
+    for (size_t i = 0; i < node_contents[0].first.size(); i++) {
+      std::array<bool, 256> hits{};
+      IndexSplitInfo info{static_cast<uint8_t>(i), 255, 0, 0};
+      for (size_t j = 0; j < node_contents.size(); j++) {
+        uint8_t v = node_contents[j].first[i];
+        if (!hits[v]) {
+          hits[v] = true;
+          info.count_++;
+          info.min_ = std::min(v, info.min_);
+          info.max_ = std::max(v, info.max_);
+        }
+      }
+      if (info.count_ > best.count_) {
+        best = info;
+      }
+    }
+    return best;
+  }
+
+  /*
+   * @param node_contents the set of key-value pairs that will be children of
+   *                      this node.
+   * @return the recursively generated tree node that leads to all of node_contents.
+   *         If there is only one entry in node_contents then a LeafNode, otherwise a BranchNode.
+   */
+  static std::unique_ptr<Node> createEqualLengthNode(std::vector<KV> node_contents) {
+    if (node_contents.size() == 1) {
+      return std::make_unique<LeafNode>(node_contents[0].first, std::move(node_contents[0].second));
+    }
+    IndexSplitInfo best = findBestSplitPoint(node_contents);
+    std::vector<std::unique_ptr<Node>> nodes;
+    nodes.resize(best.size());
+    std::sort(node_contents.begin(), node_contents.end(),
+              [index = best.index_](const KV& a, const KV& b) {
+                return a.first[index] < b.first[index];
+              });
+    auto range_start = node_contents.begin();
+    // Populate the sub-nodes for each character-branch.
+    while (range_start != node_contents.end()) {
+      // Find the first key whose character at position [best.index_] differs from the
+      // character of the current range.
+      // Everything in the range has keys with the same character at this index.
+      auto range_end = std::find_if(range_start, node_contents.end(),
+                                    [index = best.index_, c = range_start->first[best.index_]](
+                                        const KV& e) { return e.first[index] != c; });
+      std::vector<KV> next_contents;
```

Comment:
Consider adding an assert that shows convergence. Maybe something like: ASSERT(std::distance(range_start, range_end) < std::distance(node_contents.begin(), node_contents.end()))